golang/thrift/server.go

```diff
@@ -21,45 +21,106 @@ package thrift
 // THE SOFTWARE.
 
 import (
-	"reflect"
+	"io"
+	"io/ioutil"
+	"log"
+	"strings"
 
 	"github.com/apache/thrift/lib/go/thrift"
 	tchannel "github.com/uber/tchannel/golang"
 	"golang.org/x/net/context"
 )
 
-// Server is a Thrift over TChannel server.
+// Server handles incoming TChannel calls and forwards them to the matching TChanServer.
 type Server struct {
 	*tchannel.Channel
+
+	handlers map[string]TChanServer
 }
 
 // NewServer returns a server that can serve thrift services over TChannel.
 func NewServer(tchan *tchannel.Channel) *Server {
-	return &Server{tchan}
+	return &Server{
+		Channel:  tchan,
+		handlers: make(map[string]TChanServer),
+	}
 }
 
-// Register registers the given processor's methods as handlers with this server.
-// E.g., a processor 'myProcessor' with methods 'foo' and 'bar' will be registered
-// under the operations 'myProcessor::foo' and 'myProcessor::bar'.
-func (s *Server) Register(thriftName string, processorInterface reflect.Type, processor thrift.TProcessor) {
-	handler := &Handler{processor}
-	for i := 0; i < processorInterface.NumMethod(); i++ {
-		methodName := processorInterface.Method(i).Name
-		s.Channel.Register(handler, thriftName+"::"+methodName)
+// Register registers the given TChanServer to be called on any incoming call for its' services.
+// TODO(prashant): Replace Register call with this call.
+func (s *Server) Register(svr TChanServer) {
+	service := svr.Service()
+	s.handlers[service] = svr
```

Comment:
RWMutex?  I know this will typically only be called during initialization so unlikely to be invoked concurrent, but I think it's good for all methods on a given type to have the same concurrency guarantees (either all safe or all unsafe), and the overhead of an RLock is pretty low in a no-write situation.